blaze-client/src/main/scala/org/http4s/blaze/client/PoolManager.scala

```diff
@@ -201,78 +205,115 @@ private final class PoolManager[F[_], A <: Connection[F]](
-
-              case Some(pooled) =>
-                F.delay(logger.debug(s"Recycling connection for $key: $stats")) *>
-                  F.delay(callback(Right(NextConnection(pooled.conn, fresh = false))))
-
-              case None if numConnectionsCheckHolds(key) =>
-                F.delay(
-                  logger.debug(s"Active connection not found for $key. Creating new one. $stats")
-                ) *>
-                  createConnection(key, callback)
-
-              case None if maxConnectionsPerRequestKey(key) <= 0 =>
-                F.delay(callback(Left(NoConnectionAllowedException(key))))
-
-              case None if curTotal == maxTotal =>
-                val keys = idleQueues.keys
-                if (keys.nonEmpty)
-                  F.delay(
-                    logger.debug(
-                      s"No connections available for the desired key, $key. Evicting random and creating a new connection: $stats"
-                    )
-                  ) *>
-                    F.delay(keys.iterator.drop(Random.nextInt(keys.size)).next()).flatMap {
-                      randKey =>
-                        getConnectionFromQueue(randKey).map(
-                          _.fold(
-                            logger.warn(s"No connection to evict from the idleQueue for $randKey")
-                          )(_.conn.shutdown())
+  def borrow(key: RequestKey): F[NextConnection] = {
+    val connectionBorrowF =
+      F.defer(
+        Ref[F]
+          .of(false)
+          .flatMap(ref =>
+            F.async[NextConnection] { callback =>
+              semaphore.permit.use { _ =>
+                if (!isClosed) {
+                  def go(): F[Unit] =
+                    ref.get.flatMap {
+                      case true =>
+                        F.delay(
+                          logger.warn(s"Requesting connection for $key has exceed timeout")
                         ) *>
-                          decrConnection(randKey)
-                    } *>
-                    createConnection(key, callback)
-                else
-                  F.delay(
-                    logger.debug(
-                      s"No connections available for the desired key, $key. Adding to waitQueue: $stats"
-                    )
-                  ) *>
-                    addToWaitQueue(key, callback)
-
-              case None => // we're full up. Add to waiting queue.
-                F.delay(
-                  logger.debug(
-                    s"No connections available for $key.  Waiting on new connection: $stats"
-                  )
-                ) *>
-                  addToWaitQueue(key, callback)
-            }
-
-          F.delay(logger.debug(s"Requesting connection for $key: $stats")).productR(go()).as(None)
-        } else
-          F.delay(callback(Left(new IllegalStateException("Connection pool is closed")))).as(None)
-      }
+                          F.delay(callback(Left(ConnectionBorrowingException(key))))
+
```

Comment:
Is this the only reason this can get canceled?  I'm trying to reason about when we'd see this message instead of the timeout on line 311, and whether the message (and callback invocation) is accurate.

Reply:
Uh-oh, you're 100% right. Thanks for pointing that out. For sure, canceling could be caused by various reasons. But do you hint that we shouldn't invoke the callback here? I feel we can't be too sure if CE will resolve the asynchronous deadlock via the standard canceling mechanism smoothly. According to `timeoutTo` scaladoc:
> "If the source is uncancelable, the resulting effect will wait for it to complete before evaluating the fallback."

So I've tried to soften the blow.

Reply:
I'm not sure what would make that source uncancelable.  And if it is, instead of never completing, won't it just ignore the borrow?  It seems you're partially reinventing `timeoutTo`.

Reply:
After the 'cancelation in CE3' tutorial by Daniel, I'm sure I was wrong and actually was reinventing the wheel with `Ref` here, as you pointed out (if someone is interested, link to the discussion https://discord.com/channels/632277896739946517/632278585700384799/977634027651608626).
I think a few hot-takes should be highlighted. Also, this answers the question 'what makes the source uncancelable':
1. `async` is uncancelable during registration;
2. if one defines `async` without cancel token (i. e. using `None` instead of actual effect), then `async` will be uncancellable;
3. things created via `delay` is uncancellable too.

So summing up, I'm not sure that introducing timeout via `timeoutTo` will work as expected all the time. I think that for sure could be scenarios when 
1. the whole `async` effect will be considered uncancelable, 
2. and `timeoutTo` will wait for the natural execution of `async` then produce an exception (for timeout limit exceeding case). 

And I don't think that what we actually want is that behavior.

Reply:
Right.  This never returns:

```scala
IO.uncancelable(_ => IO.never)
  .timeoutTo(1.second, IO.unit)
  .unsafeRunSync()
```